### PR TITLE
Fix NONMATCHINGs and actual English Debug support

### PR DESCRIPTION
**Pull request by kazblox**
A big one. This will take long to review, but I am trying to get pokeruby up into at least a better state and parity with pokeemerald after seeing the sources unfortunately leak this early.

Most of it is multiple MATCHINGS ported from pokeemerald and pokefirered. Fixes unique to this PR however include:
- sub_80D4D64 _(new behavior? parameter poking and local temp combination)_
- **AnimTask_CreateSurfWave** _(GF used DMA3FillLarge32)_
- **_atk24_** _(permutation and bitfield representation issue)_
- CopyWallpaperTilemap, formerly sub_8099F58
- sub_8106B90 _(matched with raw pointer arithmetic instead of arrays, also parenthesis/grouping memes)_
- sub_80D3D68 _(grouping memes)_
Of which I will port over to pokeemerald/firered when merged, if compatible.

English Debug support using the ROMs from the leak is also underway.

### CODE REVIEW

**Comment by kazblox** [comment]
atk24 is now matching.

File: include/battle.h [hunk 1]
```diff
@@ -139,14 +139,14 @@ struct ProtectStruct
...
-    u8 focusBanded:1;
+    u32 statLowered:1;
+    u32 lightningRodRedirected:1;
+    u32 restoredBattlerSprite: 1;
+    u32 intimidatedMon:1;
+    u32 traced:1;
+    u32 ppNotAffectedByPressure:1;
+    u32 flag40:1;
+    u32 focusBanded:1;
     u8 field1[3];
```
Comment (PikalaxALT) [inline include/battle.h:150]: Is this padding needed?

Comment (kazblox) [reply include/battle.h:150]: Yes. The padding is needed for the struct to align.

Comment (PikalaxALT) [reply include/battle.h]: Unless the struct has `__attribute__((packed))`, filler bytes are not needed. s32 is implicitly aligned to word.

Comment (kazblox) [reply include/battle.h]: Oops. Forgot it wasn't packed. Yeah, it OK's without filler. :)

**Comment by kazblox** [comment]
I have written a WTF tier writeup on matching sub_80D4D64, for anyone who wants to see.

**Comment by kazblox** [comment]
For the record, the sub_80D4D64 trick works on Pokeblock_BufferEnhancedStatText too. Since a2 is always an uninitialized unlike sub_80D4D64's xDiff, the trickmatch statement can also be
> a2 = -a2.
Alternatively,
> a2 = 1
appears to work too.

File: sapphire_de_debug.sha1 [hunk 2]
```diff
@@ -0,0 +1 @@
+261bec8dbd7159790cd5126bd47df5b5cef61ac1  pokesapphire_de_debug.gba
```
Comment (GriffinRichards) [inline sapphire_de_debug.sha1:1]: Newline

File: src/battle/anim/water.c [hunk 3]
```diff
@@ -882,44 +673,47 @@ _080D3D32:\n\
...
+                /*
+                   It would be odd for the scanline buffers to follow ASM literal
+                   and be initialized in reverse. Experimentation based upon how
+                   compilers in general load variables and pointers showed that
+                   they were initialized this way.
+
+                   foo = bar = baz -> foo = (bar = baz)
+                */
+                gScanlineEffectRegBuffers[0][i] =
                 gScanlineEffectRegBuffers[1][i] = task->data[2];
```
Comment (GriffinRichards) [inline src/battle/anim/water.c:696]: I think breaking these assignments like this is more confusing to read than assigning them in a single line. They're not that long

File: src/battle_script_commands.c [hunk 4]
```diff
@@ -5572,288 +5572,68 @@ static void atk23_getexp(void)
...
-    if (!gBattleOutcome && (gBattleTypeFlags & BATTLE_TYPE_LINK))
+    if (gBattleControllerExecFlags == 0)
     {
-        register int found1 asm("r2");
-        register int found2 asm("r4");
-
-        //I can't for the love of god decompile that part
-
-        for (found1 = 0, i = 0; i < gBattlersCount; i += 2)
+        for (i = 0; i < 6; i++)
```
Comment (GriffinRichards) [inline src/battle_script_commands.c:5584]: `PARTY_SIZE` for these

File: src/battle_script_commands.c [hunk 5]
```diff
@@ -12852,25 +12559,29 @@ static void atkC3_trysetfutureattack(void)
...
     if (gBattleMons[gBattlerTarget].hp == 0)
+    {
         gBattlescriptCurrInstr = T1_READ_PTR(gBattlescriptCurrInstr + 1);
+    }
     else
     {
-        while (gBattleCommunication[0] < 6)
+        u8 beforeLoop = gBattleCommunication[0];
+
+        for (;gBattleCommunication[0] < 6; ++gBattleCommunication[0])
```
Comment (GriffinRichards) [inline src/battle_script_commands.c:12578]: More `PARTY_SIZE`

File: src/pokemon_storage_system_4.c [hunk 6]
```diff
@@ -2234,73 +1955,79 @@ u8 sub_809C464(void)
...
-                r4 = 0;
+                retVal = 6;
+                cursorArea = 0;
+                cursorPosition = 0;
             }
             break;
         }
-        if (JOY_NEW(A_BUTTON))
+
+        if (gMain.newKeys & A_BUTTON)
```
Comment (GriffinRichards) [inline src/pokemon_storage_system_4.c:2017]: Why drop the macros here?

File: src/battle/anim/water.c [hunk 7]
```diff
@@ -1892,17 +1301,79 @@ void sub_80D4CEC(struct Sprite *sprite)
...
+            if (xDiff) by itself won't match, so something needs to
+            be in here.
+
+            Virtually nothing what you are about to see is in the
+            assembly, you just have to to come up a strange variety
+            of compile time scenarios, question reality, and see if
+            something happens.
+        */
+        // This absolutely needs to be a negation of some kind.
+        u8 unk = -unk; // ...this is what I came up with. It matches.
```
Comment (GriffinRichards) [inline src/battle/anim/water.c:1369]: Some of these lengthier comments seem unnecessary to me, these could be abbreviated significantly

**Comment by kazblox** [comment]
@GriffinRichards Fixed.

**Pull request merged**